### PR TITLE
Create non-login users

### PR DESCRIPTION
**Pull request by mddilley**
## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/9018

## Testing
**URL to test:** 
https://md-9018-non-login-users--atd-moped-main.netlify.app/moped/

**Steps to test:**

### Creating a non-login user
1. Go to the url above and navigate to the **Staff** table
2. Click the **Add Staff** button
3. Select the **Non-login User** radio in the **Role** part of the form
4. You should see that the **Password** field disables in the form and displays a message that a password is not required for non-login users
5. Fill in the form with a first and last that you will recognize in the **Project Team tab** later on and a workgroup from the dropdown. Use your CoA email + "nonlogin" for the email field (Ex. `mike.dilley+nonlogin@austintexas.gov`). This email will be used to convert a non-login user to a Moped user in later steps.
6. Submit the form and you should see the new non-login user in the **Staff table**

### Adding a non-login user to a project team
1. Go to an existing project or create a new one
2. Go to the **Team tab** and add your new non-login user to the project

### Converting a non-login user to a Moped user w/ password
1. Go back to the **Staff table** and find your non-login user
2. Enter a valid password in the form and click the "Activate Non-login User" button
3. After submitting the form, find your user in the **Staff table** and you should see that they have the editor role and can now be inactivated
4. You should now be able to log in as the converted user as well
---
#### Ship list
- [x] Code reviewed 
- [x] Product manager approved
- [ ] Added to [QA test script](https://docs.google.com/spreadsheets/d/1n_O6MLh9cwwPf57HUM394Ea-z9uuoEV1-QW4axNZXLE/edit#) if applicable

### CODE REVIEW

File: moped-editor/src/auth/user.js [hunk 1]
```diff
@@ -294,6 +295,8 @@ export const findHighestRole = roles => {
       return "moped-editor";
     case findRole("moped-viewer"):
       return "moped-viewer";
+    case findRole(nonLoginUserRole):
+      return nonLoginUserRole;
```
Comment (mddilley) [inline moped-editor/src/auth/user.js:299]: These two lines and the import for `nonLoginUserRole` are the only changes - the rest are Prettier formatting ✨

File: moped-api/users/validation.py [hunk 2]
```diff
@@ -27,7 +27,7 @@
         "nullable": False,
         "required": True,
         "empty": False,
-        "regex": "^[a-zA-Z0-9 \-\.\']+$",
+        "regex": "^[a-zA-Z0-9 \-\.']+$",
```
Comment (mddilley) [inline moped-api/users/validation.py:30]: Black formatter removed these escapes, but they still work when I just tested on regex101.com so I guess they weren't needed in the first place. This is reminding me about [this discussion](https://github.com/cityofaustin/atd-moped/pull/740#issuecomment-1202769454) about whether we really need to validate first and last name fields (and some others)...

File: moped-api/users/validation.py [hunk 3]
```diff
@@ -78,7 +78,7 @@
         "required": True,
         "minlength": 8,
         "maxlength": 32,
-        "regex": "^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{8,}$",
+        "regex": "^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{12,}$",
```
Comment (mddilley) [inline moped-api/users/validation.py:81]: See https://github.com/cityofaustin/atd-moped/pull/789#discussion_r974479755

File: moped-editor/src/views/staff/EditStaffView.js [hunk 4]
```diff
@@ -93,6 +97,37 @@ const EditStaffView = () => {
                       validationSchema={validationSchema}
                       userCognitoId={userCognitoId}
                       isUserActive={isUserActive}
+                      FormButtons={({
+                        isSubmitting,
+                        watch,
+                        handleCloseModal,
+                        setModalState,
+                      }) => (
```
Comment (mddilley) [inline moped-editor/src/views/staff/EditStaffView.js:105]: Curious on thoughts about this pattern - there is form data that needs to be passed back up to these components from the `StaffForm` child for these buttons to work. I thought about moving the `useForm` hook up one level into the `NewStaffView` and `EditStaffView` components but that seemed overkill so this data gets passed as props when these function components are rendered in `StaffForm`.

Comment (johnclary) [reply moped-editor/src/views/staff/EditStaffView.js:119]: 👍 i think this pattern is good!

the only thing that strikes me as a bit off is seeing the `watch("someField")` props—i want those values to flow through an "onSubmit" callback that hook-form calls: `handleSubmit(onSubmit)`. but just eyeballing things quickly, that does not seem practical—there's a lot of state going on with the modal confirmations and what not.

tldr looks good to me.

File: moped-editor/src/views/staff/EditStaffView.js [hunk 5]
```diff
@@ -93,6 +105,44 @@ const EditStaffView = () => {
...
+                      }
+                      FormButtons={({
+                        isSubmitting,
+                        watch,
+                        handleCloseModal,
+                        setModalState,
+                      }) => (
+                        <>
+                          {!isNonLoginUser && (
+                            <StaffFormSaveButton disabled={isSubmitting} />
```
Comment (johnclary) [inline moped-editor/src/views/staff/EditStaffView.js:122]: @mddilley i'm guessing it's non-trivial to implement this, but it would be nice of the save button were disabled until some field value changed. 

rfh's `formState` object has an [`isdirty`](https://react-hook-form.com/api/useform/formstate) prop that you can use to determine that, but i'm not sure how easily you can get that value into this `disabled` prop, e.g. `disabled={isSubmitting || !isDirty}`. i'm happy to create a separate issue for it.

File: moped-editor/src/views/staff/test/helpers.test.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  it("transforms a role string from the form to an array the DB expects", () => {
+    const transformed = transformFormDataIntoDatabaseTypes(formOutputData);
+    const { roles } = transformed;
+
+    expect(roles.includes("moped-editor")).toBe(true);
+    expect(roles).toEqual(expect.any(Array));
+  });
+});
+
+describe("isUserNonLoginUser()", () => {
```
Comment (johnclary) [inline moped-editor/src/views/staff/test/helpers.test.js:66]: 💯  love all these tests — i am inspired!

File: moped-editor/src/views/staff/components/NonLoginUserActivationButtons.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  };
+
+  const handleActivateNonLoginUserConfirm = () => {
+    if (!passwordLooksGood(password)) {
+      setModalState({
+        open: true,
+        title: "Error",
+        message: (
+          <Typography>
+            The password is required when activating a user. It needs to be 8
```
Comment (johnclary) [inline moped-editor/src/views/staff/components/NonLoginUserActivationButtons.js:79]: It needs to be **12** characters long :)

**Review by johnclary — CHANGES_REQUESTED** [review]
@mddilley this is so cool, and unlocks really nice functionality for our users. thanks for making this happen.

i found one reference to `8` instead of `12` character passwords, but otherwise this is ready to ship 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢

File: moped-editor/src/queries/staff.js [hunk 8]
```diff
@@ -19,3 +19,42 @@ export const GET_USER = gql`
     }
   }
 `;
+
+export const ADD_NON_MOPED_USER = gql`
+  mutation AddNonMopedUser($object: moped_users_insert_input!) {
+    insert_moped_users_one(
+      object: $object
+      on_conflict: { constraint: moped_users_email_key }
```
[the remark this answers is not in the document]
Comment (mddilley) [reply moped-editor/src/queries/staff.js:27]: Great catch - this is a loose end that I forgot to tie up. Hasura does ignore the request which is great 👍 , but the user doesn't get any feedback which isn't great 👎. I'm going to think on this today and come up with a plan.

File: moped-editor/src/views/staff/EditStaffView.js (hunk 4 above)
Comment (mddilley) [reply moped-editor/src/views/staff/EditStaffView.js:119]: Thanks for the feedback! That also felt off to me. I'm really hoping to roll some updates to that modal in with https://github.com/cityofaustin/atd-data-tech/issues/10085. I think it could be possible to untangle some of the modal state/props. 🤞

File: moped-editor/src/views/staff/components/NonLoginUserActivationButtons.js (hunk 7 above)
Comment (mddilley) [reply moped-editor/src/views/staff/components/NonLoginUserActivationButtons.js]: Thanks for catching this 🙏

File: moped-editor/src/queries/staff.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (johnclary) [reply moped-editor/src/queries/staff.js:27]: sounds good. doesn't feel like a dealbreaker—we can move the UI feedback to a separate issue.

**Review by johnclary — APPROVED** [review]
🚢

File: moped-editor/src/queries/staff.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (johnclary) [reply moped-editor/src/queries/staff.js:27]: ✨ nice!

File: moped-editor/src/views/staff/EditStaffView.js (hunk 5 above)
Comment (mddilley) [reply moped-editor/src/views/staff/EditStaffView.js:122]: This would be a great improvement, and I think that you are spot on about how to implement this. Could be a good snackoo issue. 🍫

File: moped-editor/src/views/staff/NewStaffView.js [hunk 9]
```diff
@@ -49,12 +66,36 @@ const NewStaffView = () => {
...
-      method: "post",
-      path: "/users/",
-      payload: data,
-      callback,
-    });
+    const isNonLoginUser = isUserNonLoginUser(data.roles);
+
+    if (isNonLoginUser) {
+      // Remove the password for the non-login user create mutation
+      const { password, ...restOfData } = data;
```
Comment (chiaberry) [inline moped-editor/src/views/staff/NewStaffView.js:73]: nice!

File: moped-editor/src/views/staff/components/StaffUpdateUserStatusButtons.js [hunk 10]
```diff
@@ -6,23 +6,11 @@ import {
   passwordLooksGood,
   fieldParsers,
 } from "../helpers";
+import { useButtonStyles } from "./StaffFormButtons";
```
Comment (chiaberry) [inline moped-editor/src/views/staff/components/StaffUpdateUserStatusButtons.js:9]: I like this, so we dont have to repeat so much!

**Review by chiaberry — APPROVED** [review]
great! I tested and noticed that the 8 character message was still appearing, but I am guessing the test instance wasnt re-deployed after you updated that messaging. 

There isn't an option to inactivate a Non Login User, is that something we'd want to include down the line?

**Comment by mddilley** [comment]
@chiaberry Definitely, I just created https://github.com/cityofaustin/atd-data-tech/issues/10324 to cover the work of being able to inactivate a non-login user. 🙏
